### ISSUE TITLE
Performance degradation after #7593

### ISSUE THREAD

**Issue by romen** [openssl/openssl#7622]
# TL;DR

I am opening this issue to discuss about backporting performance improvements from 1.1.1 after the performance degradation introduced by #7593.

# Introduction

#7593 backported to 1.0.2 the security fixes needed to avoid the wNAF non-constant-time code path for EC scalar multiplication with generic (i.e., with no specialized implementation) prime curves.

After #7593, when dealing with secret scalars, the scalar multiplication is ultimately implemented through `ec_mul_consttime` which uses a Montgomery ladder implementation, which has the benefit of calling a fixed sequence of `EC_POINT_dbl` and `EC_POINT_add` depending only on the bit length of the scalar (which is padded to always be equal to the field size for a given elliptic curve).

# Performance loss

Directly using `EC_POINT_dbl` and `EC_POINT_add` rather than using a specialized function to perform a "ladder step" (i.e., *differential addition*) exploiting curve characteristics incurs in a considerable performance loss.

In 1.1.1, this issue was addressed by adding scaffolding for specialized implementations of the ladder step: the issue backporting that to 1.0.2 is that it requires altering some of the structs to add new function pointers in the method structures, and, given that the structures in 1.0.2 are not opaque, we might want to think twice about altering those structures to avoid obscure compatibility issues.

Until we backport the performance improvements we can expect performance degradation similar to what was reported for 1.1.1 during the beta development cycle (e.g., #6512). See next paragraph to get a quick ballpark estimate of the scale of performance loss for 1.0.2.

## A data point (ECDH on NIST P-384)

Slowing down the EC scalar multiplication will have an impact on:
- ECDH keygen and ECDSA keygen/sign (single long-term/ephemeral scalar * fixed generator point)
- ECDH key exchange (single scalar * generic point)

It does not impact ECDSA verification, as all the inputs there are public and the previous non-constant-time implementation is safe to use.

To quickly get an idea of the scale of the expected performance loss, I measured ECDH key exchange over NIST P-384 before and after #7593, recording a 37% loss, detailed output below.

### Detailed output

With previous `OpenSSL_1_0_2-stable` (at 59b9c67fcaf1c1e2c0e30de6facca85910ac361a):

~~~
OPENSSL_CONF=apps/openssl.cnf util/opensslwrap.sh speed ecdhp384
Doing 384 bit  ecdh's for 10s: 7466 384-bit ECDH ops in 9.98s
OpenSSL 1.0.2q-dev  xx XXX xxxx
built on: reproducible build, date unspecified
options:bn(64,64) md2(int) rc4(16x,int) des(idx,cisc,16,int) aes(partial) idea(int) blowfish(idx)
compiler: gcc -I. -I.. -I../include  -DOPENSSL_THREADS -D_REENTRANT -DDSO_DLFCN -DHAVE_DLFCN_H -Wa,--noexecstack -DBN_DEBUG -DREF_CHECK -DCONF_DEBUG -DCRYPTO_MDEBUG -m64 -DL_ENDIAN -g -Wall -DOPENSSL_IA32_SSE2 -DOPENSSL_BN_ASM_MONT -DOPENSSL_BN_ASM_MONT5 -DOPENSSL_BN_ASM_GF2m -DRC4_ASM -DSHA1_ASM -DSHA256_ASM -DSHA512_ASM -DMD5_ASM -DAES_ASM -DVPAES_ASM -DBSAES_ASM -DWHIRLPOOL_ASM -DGHASH_ASM -DECP_NISTZ256_ASM -pedantic -DPEDANTIC -Wno-long-long -Wsign-compare -Wmissing-prototypes -Wshadow -Wformat -Wundef -Werror -DCRYPTO_MDEBUG_ALL -DCRYPTO_MDEBUG_ABORT -DOPENSSL_NO_DEPRECATED                                 
                              op      op/s
 384 bit ecdh (nistp384)   0.0013s    748.1
~~~

After #7593 (at b18162a7c9bbfb57112459a4d6631fa258fd8c0c):

~~~
OPENSSL_CONF=apps/openssl.cnf util/opensslwrap.sh speed ecdhp384
Doing 384 bit  ecdh's for 10s: 4707 384-bit ECDH ops in 9.98s
OpenSSL 1.0.2q-dev  xx XXX xxxx
built on: reproducible build, date unspecified
options:bn(64,64) md2(int) rc4(16x,int) des(idx,cisc,16,int) aes(partial) idea(int) blowfish(idx)
compiler: gcc -I. -I.. -I../include  -DOPENSSL_THREADS -D_REENTRANT -DDSO_DLFCN -DHAVE_DLFCN_H -Wa,--noexecstack -DBN_DEBUG -DREF_CHECK -DCONF_DEBUG -DCRYPTO_MDEBUG -m64 -DL_ENDIAN -g -Wall -DOPENSSL_IA32_SSE2 -DOPENSSL_BN_ASM_MONT -DOPENSSL_BN_ASM_MONT5 -DOPENSSL_BN_ASM_GF2m -DRC4_ASM -DSHA1_ASM -DSHA256_ASM -DSHA512_ASM -DMD5_ASM -DAES_ASM -DVPAES_ASM -DBSAES_ASM -DWHIRLPOOL_ASM -DGHASH_ASM -DECP_NISTZ256_ASM -pedantic -DPEDANTIC -Wno-long-long -Wsign-compare -Wmissing-prototypes -Wshadow -Wformat -Wundef -Werror -DCRYPTO_MDEBUG_ALL -DCRYPTO_MDEBUG_ABORT -DOPENSSL_NO_DEPRECATED                                 
                              op      op/s
 384 bit ecdh (nistp384)   0.0021s    471.6
~~~

# A potential plan

At some point in the future (right now I don't have enough time available), I plan to open a separate PR to backport the following:
- #6690 (probably only the scaffolding, not the Lopez-Dahab implementation for binary curves)
- #6772 + #7000 (specialized ladder implementation for prime curves, the additional tests are unfortunately hard to backport)

# More discussion points

## The expected gain will only partially fill the performance gap

We must consider that following the above backporting plan will not deliver the same performance that we had before #7593 : take as a reference the numbers reported in #6772 (which are slightly worsened by #7000). From there we can see that we can gain around a 20-25% boost between the unoptimized and the optimized ladder.

## About binary curves

We might want to discuss if we really want to additionally backport #6070 (to unify the scalar multiplication path for prime and binary curves) and then the Lopez-Dahab specialized implementation from #6690.
I, for one, am not sure it's worth to do it, as the worse testing coverage of 1.0.2 carries an heightened risk of unnoticed bugs and compatibility issues with user code.

## More in the agenda

Note to self: verify what is the status about backport to 1.0.2 of #6116 for future work.

**Comment by romen** [comment]
Ping @mattcaswell and @paulidale (that followed this discussion in #7593 ) and @bbbrumley  (if he wants to weight in this discussion).

**Comment by mattcaswell** [comment]
> add new function pointers in the method structures

Generally speaking changing the size of a transparent struct is not allowed in an ABI compatible release. If a transparent struct is stack allocated in an application, and the library version is upgraded without recompiling the application then a failure could occur because of the different size of the struct. This is one of the major motivating factors that made us move to opaque structs in the first place.

**Comment by romen** [comment]
@mattcaswell I wish there was a :facepalm: reaction in Github!

Of course you are right and I should have thought of that from the beginning. 

One potential solution for this problem is then implementing a `ec_scalar_mul_ladder_ex` function in `ec_mul.c` which behaves like 1.1.1 `ec_scalar_mul_ladder` but takes as extra parameters the ladder scaffold function pointers.
Then in `ecp_meth.c` we could define a custom `mul` implementation that conditionally calls the wNAF implementation or `ec_scalar_mul_ladder_ex` with the specialized pointers depending on the number of inputs (we will have to duplicate the initial part of the wNAF `ec_mul` where the same check is done).

**Comment by mattcaswell** [comment]
Try `:man_facepalming:` :man_facepalming: :smile:

It all starts to sound like quite a significant change to be making to a stable branch.

**Comment by mattcaswell** [comment]
Or `:woman_facepalming:` works too...:woman_facepalming:

**Comment by bbbrumley** [comment]
My 2c, even though I don't think my vote counts: I do admire @romen's ambition, but 1.0.2 is security fix only mode.

We don't have as extensive regression testing for 1.0.2. And who knows what FrankenVersion downstream has with selective patches. I feel like: somewhat OK performance recovery, vs security risk of us mucking something up and anyway a lot of effort.

Plan B would be let researchers kick the tires, and cross the bridge when we come to it  :man_facepalming: :woman_facepalming:

**Comment by mattcaswell** [comment]
> 1.0.2 is security fix only mode.

Technically it isn't - yet. Security fix only mode for 1.0.2 starts on 1st January 2019.

Still the lack of testing in 1.0.2 for what would be a significant change is a big concern.

**Comment by paulidale** [comment]
My leaning would be to avoid back porting this and live with the performance loss.  It sounds invasive, we've got to be careful with structures and we're concerned about the lack of test coverage.

I'm also not sufficiently invested to say _no_ to a back port if one were presented.

**Comment by mspncp** [comment]
I noticed we even have a label for this case... :-)
